clang/include/clang/CIR/Interfaces/ASTAttrInterfaces.td

```diff
@@ -0,0 +1,213 @@
+    > {
+        let methods = [
+            InterfaceMethod< "", "bool", "isCopyAssignmentOperator", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    if (auto decl = dyn_cast< clang::CXXMethodDecl >($_attr.getAstDecl()))
+                        return decl->isCopyAssignmentOperator();
+                    return false;
+                }]
+            >,
+            InterfaceMethod< "", "bool", "isMoveAssignmentOperator", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    if (auto decl = dyn_cast< clang::CXXMethodDecl >($_attr.getAstDecl()))
+                        return decl->isMoveAssignmentOperator();
+                    return false;
+                }]
+            >,
+            InterfaceMethod< "", "bool", "isConst", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->isConst();
+                }]
+            >
+        ];
+    }
+
+    def ASTCXXConstructorDeclInterface : AttrInterface< "ASTCXXConstructorDeclInterface",
+        [ASTCXXMethodDeclInterface]
+    > {
+        let methods = [
+            InterfaceMethod< "", "bool", "isDefaultConstructor", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->isDefaultConstructor();
+                }]
+            >,
+            InterfaceMethod< "", "bool", "isCopyConstructor", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->isCopyConstructor();
+                }]
+            >
+        ];
+    }
+
+    def ASTCXXConversionDeclInterface : AttrInterface< "ASTCXXConversionDeclInterface",
+        [ASTCXXMethodDeclInterface]
+    > {}
+
+    def ASTCXXDestructorDeclInterface : AttrInterface< "ASTCXXDestructorDeclInterface",
+        [ASTCXXMethodDeclInterface]
+    > {}
+
+    def ASTTypeDeclInterface : AttrInterface< "ASTTypeDeclInterface",
+        [ASTNamedDeclInterface]
+    > {}
+
+    def ASTTagDeclInterface : AttrInterface< "ASTTagDeclInterface",
+        [ASTTypeDeclInterface, ASTDeclContextInterface]
+    > {
+        let methods = [
+            InterfaceMethod< "", "clang::TagTypeKind", "getTagKind", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->getTagKind();
+                }]
+            >
+        ];
+    }
+
+    def ASTEnumDeclInterface : AttrInterface< "ASTEnumDeclInterface",
+        [ASTTagDeclInterface]
+    > {}
+
+    def ASTRecordDeclInterface : AttrInterface< "ASTRecordDeclInterface",
+        [ASTTagDeclInterface]
+    > {}
+
+    def ASTCXXRecordDeclInterface : AttrInterface< "ASTCXXRecordDeclInterface",
+        [ASTRecordDeclInterface]
+    > {
+        let methods = [
+            InterfaceMethod< "", "bool", "isLambda", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->isLambda();
```

Comment:
Thanks for mapping at length all the stuff we currently use from the AST. As we move into encoding some extra information directly into CIR, I believe we might slowly get rid of some of these, and this will naturally serve as documentation of how much AST we end up using all around.

Is there any nodes you've added for the sake of completeness or is all of this actually used in CIR? I'd suggest we keep this list only reflecting what we currently use, so we don't unnecessarily bloat out the interface with things we don't yet have a use case for.

Reply:
This was everything I needed to compile and pass tests. I only added additional intermediate Interfaces from AST hierarchy, which are empty in such case. Also to replicate API of ASTs I put methods to corresponding Interfaces, even though you do not use the interface directly, but inherit the method from base interface, e.g. `NamedDecl` provide `getDeclName` which you use only through `VarDeclInterface`.